### PR TITLE
refactor(core): add profile verification

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

Add profile verification method to verify the provided profile payload:

1. For all the protected identifiers, email, phone, and social, a verified identifier record must present
2. Profile fulfilling, a valid user account must be present in the identifier record.
3. Profile fulfilling, all the field must not exist
4. New register profile payload must contain valid identifiers to create a new account.  i.e. (a single username is invalid)
5. New registered and profile-fulfilling fields must not be registered by other user accounts.
6. ForgotPassord profile, new password must not equal the old one.

Please ignore the huge PR size, mostly are test cases. 

## Testing

UT case added

### CODE REVIEW

File: packages/core/src/routes/interaction/verifications/profile-verification.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    new RequestError({
+      code: 'session.unauthorized',
+      status: 401,
+    })
+  );
+
+  return findUserById(accountIdentifier.value);
+};
+
+const protectedIdentifierVerification = (
```
Comment (charIeszhao) [inline packages/core/src/routes/interaction/verifications/profile-verification.ts:41]: The function name should be in a verb-object format. For example, `verifyProtectedIdentifier()`

File: packages/core/src/routes/interaction/verifications/profile-verification.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+  { email, phone, connectorId }: Profile,
+  identifiers: Identifier[]
+) => {
+  if (email) {
+    assertThat(
+      identifiers.some(({ key, value }) => key === 'verifiedEmail' && value === email),
```
Comment (charIeszhao) [inline packages/core/src/routes/interaction/verifications/profile-verification.ts:47]: Maybe we should use `emailVerified` in order to be consistent with OIDC terminology?

File: packages/core/src/routes/interaction/verifications/profile-verification.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+  const { profile, event } = ctx.interactionPayload;
+
+  if (!profile) {
+    return;
+  }
+
+  protectedIdentifierVerification(profile, identifiers);
+
+  if (event === Event.SignIn) {
+    // Find exist account
```
Comment (charIeszhao) [inline packages/core/src/routes/interaction/verifications/profile-verification.ts:190]: nit: existing

File: packages/core/src/routes/interaction/verifications/profile-verification.ts (hunk 1 above)
Comment (simeng-li) [reply packages/core/src/routes/interaction/verifications/profile-verification.ts]: udpated

File: packages/core/src/routes/interaction/verifications/profile-verification.ts (hunk 2 above)
Comment (simeng-li) [reply packages/core/src/routes/interaction/verifications/profile-verification.ts]: udpated

File: packages/core/src/routes/interaction/verifications/profile-verification.ts (hunk 3 above)
Comment (simeng-li) [reply packages/core/src/routes/interaction/verifications/profile-verification.ts]: udpated

**Review by charIeszhao — APPROVED** [review]

**Pull request merged**